### PR TITLE
Library data requirements

### PR DESCRIPTION
**Pull request by natjoe4**
# Summary
Measure repository service now supports Library $data-requirements operations!

## New behavior
The Library $data-requirements endpoint is now exposed and working.

## Code changes
 - exposed `Library/$data-requirements` and `Library/:id/$data-requirements` endpoints in serverConfig
 - updated request schemas to fix a message bug and include Library $data-requirements schema
 - updated Library service to include handling for Library $data-requirements
 - updated `createLibraryPackage` function to take optional `rootLibRef` param to pass into fqm-execution `calculateLibraryDataRequirementsFunction`
 - updated Readme to reflect Library $data-requirements functionality
 - updated capabilityStatement to reflect Library $data-requirements. Also reformatted the JSON file to be more in line with package.lock 
 - Changed fqm-execution dependency to point to Github since we have not yet released a new version with necessary changes
 - Added testing

# Testing guidance
 - `npm install`
 - `npm run db:reset`
 - Head into `ecqm-content-r4-2021/bundles/measure/ColorectalCancerScreeningsFHIR/ColorectalCancerScreeningsFHIR-bundle.json` and find the `Library` resource in the bundle with `"id": "ColorectalCancerScreeningsFHIR"`. Add the following code snippet to the `Library`: 
 ```
        "identifier": {
          "system": "testSystem",
          "value": "testCode"
        }
```
 - run `npm run db:loadBundle "ecqm-content-r4-2021/bundles/measure/ColorectalCancerScreeningsFHIR/ColorectalCancerScreeningsFHIR-bundle.json"`
 - Now use the attached Insomnia bundle to test all requests
 - Feel free to reach out with any questions!
[library-$data-requirements.json.zip](https://github.com/projecttacoma/measure-repository-service/files/10724499/library-.data-requirements.json.zip)

### CODE REVIEW

File: src/config/serverConfig.ts [hunk 1]
```diff
@@ -111,6 +111,30 @@ export const serverConfig: ServerConfig = {
...
+          name: 'dataRequirements',
+          route: '/$data-requirements',
+          method: 'GET',
+          reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Measure-data-requirements'
+        },
+        {
+          name: 'dataRequirements',
+          route: '/$data-requirements',
+          method: 'POST',
+          reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Measure-data-requirements'
```
Comment (sarahmcdougall) [inline src/config/serverConfig.ts:125]:
```suggestion
          reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Library-data-requirements'
```

File: src/config/serverConfig.ts [hunk 2]
```diff
@@ -111,6 +111,30 @@ export const serverConfig: ServerConfig = {
...
+        },
+        {
+          name: 'dataRequirements',
+          route: '/:id/$data-requirements',
+          method: 'GET',
+          reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Measure-data-requirements'
```
Comment (sarahmcdougall) [inline src/config/serverConfig.ts:131]:
```suggestion
          reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Library-data-requirements'
```

File: src/config/serverConfig.ts [hunk 3]
```diff
@@ -111,6 +111,30 @@ export const serverConfig: ServerConfig = {
...
+        },
+        {
+          name: 'dataRequirements',
+          route: '/:id/$data-requirements',
+          method: 'POST',
+          reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Measure-data-requirements'
```
Comment (sarahmcdougall) [inline src/config/serverConfig.ts:137]:
```suggestion
          reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Library-data-requirements'
```

File: src/config/serverConfig.ts [hunk 4]
```diff
@@ -111,6 +111,30 @@ export const serverConfig: ServerConfig = {
           route: '/:id/$package',
           method: 'POST',
           reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Library-package'
+        },
+        {
+          name: 'dataRequirements',
+          route: '/$data-requirements',
+          method: 'GET',
+          reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Measure-data-requirements'
```
Comment (sarahmcdougall) [inline src/config/serverConfig.ts:119]:
```suggestion
          reference: 'http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Library-data-requirements'
```

File: src/requestSchemas.ts [hunk 5]
```diff
@@ -90,9 +90,7 @@ const stringToBool = z
   .union([z.enum(['true', 'false']), z.boolean()])
   .transform(x => (typeof x === 'boolean' ? x : x === 'true'));
 const stringToNumber = z.coerce.number();
-const checkDate = (paramName: string) => {
-  return z.string().regex(DATE_REGEX, `${paramName} parameter is not a valid FHIR date`);
-};
+const checkDate = z.string().regex(DATE_REGEX, `Invalid FHIR date`);
```
Comment (sarahmcdougall) [inline src/requestSchemas.ts:93]:
```suggestion
const checkDate = z.string().regex(DATE_REGEX, 'Invalid FHIR date');
```

File: src/services/LibraryService.ts [hunk 6]
```diff
@@ -56,6 +57,34 @@ export class LibraryService implements Service<fhir4.Library> {
...
+    validateParamIdSource(req.params.id, params.id);
+    const query = extractIdentificationForQuery(args, params);
+    const parsedParams = parseRequestSchema({ ...params, ...query }, LibraryDataRequirementsArgs);
+
+    logger.info(`${req.method} ${req.path}`);
+
+    const { libraryBundle, rootLibRef } = await createLibraryPackageBundle(query, parsedParams);
+
+    const { results } = await Calculator.calculateLibraryDataRequirements(libraryBundle, {
+      ...(params.periodStart && { measurementPeriodStart: params.periodStart }),
```
Comment (sarahmcdougall) [inline src/services/LibraryService.ts:82]: Is there a reason why we use `params` for `calculateLibraryDataRequirements` but we use `parsedParams` when we calculate the data requirements in the `MeasureService`? Or should it also be `parsedParams` here?

**Review by sarahmcdougall — CHANGES_REQUESTED** [review]

File: src/services/LibraryService.ts (hunk 6 above)
Comment (natjoe4) [reply src/services/LibraryService.ts:82]: certainly `parsedParams` good catch!

File: test/util/bundleUtils.test.ts [hunk 7]
```diff
@@ -259,14 +268,29 @@ describe('bundleUtils', () => {
...
+      const { rootLibRef } = await createLibraryPackageBundle({ id: 'LibraryWithDeps' }, {});
+      expect(rootLibRef).toEqual('http://example.com/LibraryWithDeps|0.0.1-test');
+    });
+
+    it('returns rootLibRef with just url when url present and version missing', async () => {
+      const { rootLibRef } = await createLibraryPackageBundle({ id: 'LibWithNoDeps' }, {});
+      expect(rootLibRef).toEqual('http://example.com/LibraryWithNoDeps');
+    });
+
+    it('returns rootLibRef with just url when url present and version missing', async () => {
```
Comment (lmd59) [inline test/util/bundleUtils.test.ts:290]: I think the test description here is a bit off. Maybe:
```suggestion
    it('returns rootLibRef with id when url and version missing', async () => {
```

**Review by lmd59 — COMMENTED** [review]
Sarah's comments look good. Added one more!

**Review by lmd59 — APPROVED** [review]

**Review by sarahmcdougall — APPROVED** [review]
Great work!

**Pull request merged**